flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/PulsarWriter.java

```diff
@@ -0,0 +1,293 @@
+                                    "Failed to send message to Pulsar");
+                        } else {
+                            LOG.debug("Sent message to Pulsar {} with message id {}", topic, id);
+                        }
+                    });
+        }
+    }
+
+    private TypedMessageBuilder<IN> createMessageBuilder(
+            String topic, IN element, PulsarSinkContextAdapter context) {
+        TypedMessageBuilder<IN> builder =
+                producerRegister.createMessageBuilder(topic, element, schema);
+
+        byte[] orderingKey = serializationSchema.orderingKey(element, context);
+        if (orderingKey != null) {
+            builder.orderingKey(orderingKey);
+        }
+
+        String key = serializationSchema.key(element, context);
+        if (!Strings.isNullOrEmpty(key)) {
+            builder.key(key);
+        }
+
+        Map<String, String> properties = serializationSchema.properties(element, context);
+        if (properties != null) {
+            builder.properties(properties);
+        }
+
+        Long timestamp = serializationSchema.eventTime(element, context);
+        if (timestamp != null) {
+            builder.eventTime(timestamp);
+        }
+
+        List<String> clusters = serializationSchema.replicationClusters(element, context);
+        if (clusters != null) {
+            builder.replicationClusters(clusters);
+        }
+
+        if (serializationSchema.disableReplication(element, context)) {
+            builder.disableReplication();
+        }
+
+        // Message Delay
+        long deliverAt = messageDelayer.deliverAt(element, context);
+        if (deliverAt > 0) {
+            builder.deliverAt(deliverAt);
+        }
+
+        return builder;
+    }
+
+    @Override
+    public void flush(boolean endOfInput) throws IOException {
+        if (endOfInput) {
+            // Try flush only once when we meet the end of the input.
+            producerRegister.flush();
+        } else {
+            while (pendingMessages.availablePermits() < sinkConfiguration.getMaxPendingMessages()) {
+                producerRegister.flush();
+            }
+        }
+    }
+
+    @Override
+    public Collection<PulsarCommittable> prepareCommit() {
+        if (deliveryGuarantee == DeliveryGuarantee.EXACTLY_ONCE) {
+            return producerRegister.prepareCommit();
+        } else {
+            return emptyList();
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        // Close all the resources and throw the exception at last.
+        closeAll(metadataListener, producerRegister);
+        closed = true;
+    }
+
+    private void setupFlinkMetrics(PulsarSinkWriterMetrics pulsarSinkWriterMetrics) {
+        registerGauges();
+        registerMetricUpdateTimer();
+    }
+
+    private void registerGauges() {
+        pulsarSinkWriterMetrics.setCurrentSendTimeGauge(this::getCurrentSendTime);
+        producerRegister.registerProducerGauges();
```

Comment:
Producer specific metrics should be registered upon producer creation, not at startup.